src/models/catboost.py

```diff
@@ -42,64 +90,171 @@ def load_fold(
+
+        for idx, sample in enumerate(samples):
+            target_abstract = sample.get("abstract", "")
+            target_tokens = simple_tokenizer(target_abstract)
+            author_papers = sample["author"]["papers"]
+            author_scores = []
+            
+            for paper in author_papers:
+                doc_abstract = paper.get("abstract", "")
+                doc_tokens = simple_tokenizer(doc_abstract)
+                if doc_tokens:
+                    # Compute BM25 score between target and author paper using compute_bm25_score
+                    score = self.compute_bm25_score(target_tokens, doc_tokens)
+                    author_scores.append(score)
+
+            if author_scores:
+                bm25_max_scores.append(max(author_scores))
+                bm25_avg_scores.append(sum(author_scores) / len(author_scores))
+            else:
+                bm25_max_scores.append(0.0)
+                bm25_avg_scores.append(0.0)
+
+        X = X.copy()  # Make a copy to avoid modifying the original DataFrame
+        # Add the BM25 scores as new columns
+        X["bm25_max_score"] = bm25_max_scores
+        X["bm25_avg_score"] = bm25_avg_scores
+        return X
+
+    def fit(self, data: Data):
+        """
+        1. Build the BM25 corpus and initialize the BM25 model.
+        2. Load and process the training data.
+        3. Fit the preprocessing pipeline on the training data.
+        4. Process validation data.
+        5. Train the model on the processed training data.
         """
-1. Fit the preprocessing pipeline on the training data
-2. Convert the training and validation data using the preprocessing pipeline
-3. Train the model on the processed training and validation data
-"""
+        # Build the BM25 model using the training data
+        corpus_documents = []
+        for sample in tqdm(data.get_fold("train"), desc="Building BM25 corpus"):
+            target_abstract = sample.get("abstract", "")
+            target_tokens = simple_tokenizer(target_abstract)
+            if target_tokens:
+                corpus_documents.append(target_tokens)
+            for paper in sample["author"]["papers"]:
+                paper_abstract = paper.get("abstract", "") or ""
+                paper_tokens = simple_tokenizer(paper_abstract)
+                if paper_tokens:
+                    corpus_documents.append(paper_tokens)
+
+        # Initialize BM25 model with the training corpus
+        self.bm25_model = BM25Okapi(corpus_documents)
+        print("BM25 model initialized on training corpus.")
+
+        # Load the training data
         X_train, y_train = self.load_fold(data, "train")
-        self.feature_processing_pipeline = ColumnTransformer([
-            ('passthrough', 'passthrough', ["referenceCount", "author_num_papers", "is_cited"]),
-            ("paper_categories", CountVectorizer(analyzer=passthrough_func), "categories"),
+        samples_train = data.get_fold("train")

```

Comment:
You shouldn't call `get_fold`. You should use the outputs of `load_fold` (and add missing information if you need it)